### ISSUE TITLE
Investigate the edge pixel in the direct method

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#217]
Following the discussion is https://github.com/PyAbel/PyAbel/pull/214#issuecomment-407919764, we may want to think about how to deal with the last element of the output. Currently, this is strictly zero, which makes some degree of sense, because a basic assumption of the Abel transform is that the intensity goes to zero at the last pixel. However, other transform methods (at least the hansenlaw) allow this pixel to have non-zero values. Potentially we could pad the array by an additional value (either zero, or the same as the last pixel) and then trim the final array.

**Comment by stggh** [comment]
> @stggh, do you agree?

No. Unless I have misinterpreted this discussion. Zero-padding will not repair a faulty image (that has a background intensity). This figure show the inverse able transform radial profile for the 'O-' test image, including an added background intensity, with and without a zero-intensity boundary. The background results in an additional structure (~500 pixels). This pseudo structure occurs for the other methods, except `hansenlaw`:
     
![Image](https://github.com/user-attachments/assets/23c2fe31-7785-4d30-911a-9e5b4a81bfbb)

Code:
```python
import abel
import matplotlib.pyplot as plt

fn = 'O-'
aim = abel.tools.analytical.SampleImage(n:=1001, name=fn).abel
print(f'Image intensity maximum = {aim.max():.0f}')

# pseudo backround
aim5 = aim + 5

# zero boundary
aim50 = aim5
aim50[:, 0] = aim50[:, -1] = 0

# inverse Abel transform -------------
method = 'direct'
im = abel.Transform(aim, method=method, angular_integration=True)
im5 = abel.Transform(aim5, method=method, angular_integration=True)
im50 = abel.Transform(aim50, method=method, angular_integration=True)

# plot -----------
plt.plot(*im.angular_integration,label='no background')
plt.plot(*im5.angular_integration, label='background 5')
plt.plot(*im50.angular_integration, '--', label='background 5, 0 edge')

plt.legend()
plt.title(method)
plt.xlabel('radius')
plt.ylabel('amplitude')
plt.savefig('bkg.svg')
plt.show()
```

**Comment by MikhailRyazanov** [comment]
I was mostly concerned about the following behavior when any noise at the edge pixel shifts the whole transform intensity (demonstrated using the low-level 1D transforms for clarity):
<details>
<summary>[code for the plots]</summary>

```python
import abel
import matplotlib.pyplot as plt

aim = abel.tools.analytical.GaussianAnalytical(50, 5, symmetric=False).abel

# add "noise" at edge
aime1 = aim.copy()
aime1[-1] = 1

# add "noise" just before edge
aime0 = aim.copy()
aime0[-2] = 1

fig = plt.figure(figsize=(6, 8))

plt.subplot(3, 1, 1)
plt.title('Inputs')
plt.xlabel('radius')
plt.ylabel('amplitude')

plt.plot(aim, c='C4', label='clean')
plt.plot(aime1, ':', c='C4', label='edge=1')
plt.plot(aime0, '.--', c='C4', label='edge=0')

plt.legend()


plt.subplot(3, 1, (2, 3))
plt.title('Transforms')
plt.xlabel('radius')
plt.ylabel('amplitude')

for c, method in enumerate(['direct', 'hansenlaw', 'daun']):
    transform = eval(f'abel.{method}.{method}_transform')
    im = transform(aim)
    ime1 = transform(aime1)
    ime0 = transform(aime0)

    plt.plot(im, c=f'C{c}', label=f'{method} clean')
    plt.plot(ime1, ':', c=f'C{c}', label=f'{method} edge=1')
    plt.plot(ime0, '.--', c=f'C{c}', label=f'{method} edge=0')

plt.legend()

plt.tight_layout()
plt.savefig('edge.svg')
plt.show()
```
</details>

![Image](https://github.com/user-attachments/assets/aaa76e98-9801-4889-88dc-4dcca69d8c39)

* All methods are consistent for the “clean” Gaussian (solid curves; a minor difference among the methods can be seen only at small radii, otherwise they overlap, as expected).
* When the edge pixel is disturbed (“edge=1”, dotted line), all “other” methods, represented by Daun, show some expected noise near that pixel, but it does not propagate much inwards. However, for Direct and Hansel–Law, this *shifts the whole* transformed intensity.
* The same kind of disturbance but not at the very edge (“edge=0”, dashed line with circles) causes a similar expected transform noise for all the methods, but without such overall shifts even for Direct and Hansel–Law.

So basically my point is that the behavior of Direct and Hansel–Law in the “edge=1” case is not consistent with other methods and, most likely, not expected by and not desirable to many users. And therefore, it would be useful to add an option to treat the edge pixel indeed as a pixel (after which the intensity is zero everywhere) instead of implicitly extrapolating its value to infinity (and then effectively subtracting that “background” from the whole input).

**Comment by MikhailRyazanov** [comment]
Regarding your code. Here:
```python
# zero boundary
aim50 = aim5
aim50[:, 0] = aim50[:, -1] = 0
```
it must be ``aim50 = aim5.copy()`` because otherwise ``aim50`` is just an alias to ``aim5``, and thus ``aim50[:, 0] = aim50[:, -1] = 0`` also changes ``aim5`` the same way. If you redo the plots with ``.copy()``, they will show that “background 5” is the same as “no background”, but “background 5, 0 edge” indeed has some background. And using ``method = 'hansenlaw'`` shows the same picture, but all other methods are similar to what your original plot showed (have background regardless of the edge value).

**Comment by MikhailRyazanov** [comment]
> This is pretty neat, but an extreme example.

The transforms are linear, so the effect is just directly proportional to that pixel value. For real data it will be smaller but still cause a systematic shift.

> The image intensity is required to decay to zero more rapidly than `1/r`, these edge cases would appear to violate this condition?

True, and we in fact say that intensity must be actually 0 outside. But most methods consider that the outermost pixel has some meaningful data, and the intensity drops to zero immediately after it. In contrast, Direct and Hansel–Law think that the outermost pixel intensity extends to infinity (which would indeed violate the `1/r` condition), and thus the user must set it to zero explicitly to get meaningful transform results. So the idea here is to add an implicit zero-valued “pixel” just outside the data range, allowing Direct and Hansel–Law to use the whole data range without being confused by the edge value.

**Comment by MikhailRyazanov** [comment]
> Will 2-zero pixel values need to be added, to maintain the 'image' odd-pixel number requirement?

This should be done for the low-level functions (``<method>_transform()``), which work on half-images and thus don’t care about the size parity. Moreover, in principle, we don’t actually need to add and then strip that extra pixel. The desired effect can be achieved by adjusting the derivative value at the outer edge (in theory; I haven’t investigated yet what exactly needs to be done).

**Comment by MikhailRyazanov** [comment]
The original issue was resolved in #419, and now #420 resolved a similar for the Hansen–Law method, so this can be safely closed.
